_posts/2024-04-12-vector-similarity-search-vss.md

```diff
@@ -0,0 +1,127 @@
+
+```sql
+SELECT *
+FROM embeddings
+ORDER BY array_distance(vec, [1,2,3]::FLOAT[3])
+LIMIT 3;
+```
+
+will have their logical plan optimized to become a projection over a new `HNSW` index scan operator, removing the limit and sort altogether. We can verify this by checking the `EXPLAIN` output:
+
+```sql
+EXPLAIN
+SELECT *
+FROM embeddings
+ORDER BY array_distance(vec, [1,2,3]::FLOAT[3])
+LIMIT 3;
+```
+
+```text
+┌───────────────────────────┐
+│         PROJECTION        │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│             #0            │
+└─────────────┬─────────────┘
+┌─────────────┴─────────────┐
+│         PROJECTION        │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│            vec            │
+│array_distance(vec, [1.0, 2│
+│         .0, 3.0])         │
+└─────────────┬─────────────┘
+┌─────────────┴─────────────┐
+│      HNSW_INDEX_SCAN      │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│   t1 (HNSW INDEX SCAN :   │
+│            idx)           │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│            vec            │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│           EC: 3           │
+└───────────────────────────┘
+```
+
+You can pass the `HNSW` index creation statement a `metric` parameter to decide what kind of distance metric to use. The supported metrics are `l2sq`, `cosine` and `inner_product`, matching the three built-in distance functions: `array_distance`, `array_cosine_distance` and `array_inner_product`.
+The default is `l2sq`, which uses Euclidean distance (`array_distance`):
+
+```sql
+CREATE INDEX l2sq_idx
+ON embeddings
+USING HNSW (vec)
+WITH (metric = 'l2sq');
+```
+
+To use cosine distance (`array_cosine_distance`):
+
+```sql
+CREATE INDEX cos_idx
+ON embeddings
+USING HNSW (vec)
+WITH (metric = 'cosine');
+```
+
+To use inner product (`array_inner_product`):
+
+```sql
+CREATE INDEX ip_idx
+ON embeddings
+USING HNSW (vec)
+WITH (metric = 'ip');
+```
+
+## Limitations
+
+It is important to know that the `HNSW` index is an _approximate_ index, meaning that the results returned by the index may not be the _exact_ top-n results, but rather a set of results that are _close_ to the top-n results. The `HNSW` in `HNSW` index stands for Hierarchical Navigable Small World, which is a graph-based index structure that is particularly well-suited for high-dimensional vector similarity search.
+
+While the `HNSW` index supports insertions and deletes to the base table after the index has been created, deletes are not immediately reflected internally in the index structure, but are instead marked as deleted, and will only be removed after a certain number of updates have been made. This is a trade-off to maintain the performance of updates, but will lead to the _quality_ of the index deteriorating over time, impacting both speed and accuracy of retrieval. To remedy this you can use the pragma `pragma_hnsw_compact_index('<index name>')` to trigger a recompaction, which will be faster than dropping and recreating the index.
```

Comment:
I noticed that we use the pragma function here, rather than the pragma call, as we do in the `vss` docs:
``PRAGMA hnsw_compact_index('<index name>')``

Maybe we want to pick one of them, just to avoid needless confusion